promql/query_logger_test.go

```diff
@@ -163,3 +171,65 @@ func TestParseBrokenJSON(t *testing.T) {
 		})
 	}
 }
+
+// Closing ActiveQueryTracker instance shouldn't cause panic if the engine still tries to log queries.
+func TestActiveQueryTrackerClose(t *testing.T) {
+	dir := t.TempDir()
+
+	maxConcurrent := 10
+	fileAsBytes, closer, err := getMMappedFile(path.Join(dir, "queries.active"), 1+maxConcurrent*entrySize, nil)
+	require.NoError(t, err)
+
+	isRunning := &atomic.Bool{}
+	isRunning.Store(true)
+
+	queryLogger := ActiveQueryTracker{
+		isRunning:     isRunning,
+		maxConcurrent: maxConcurrent,
+		mmappedFile:   fileAsBytes,
+		closer:        closer,
+		logger:        nil,
+		getNextIndex:  make(chan int, 4),
+	}
+
+	queryLogger.generateIndices(4)
+	veryLongString := "MassiveQueryThatNeverEndsAndExceedsTwoHundredBytesWhichIsTheSizeOfEntrySizeAndShouldThusBeTruncatedAndIamJustGoingToRepeatTheSameCharactersAgainProbablyBecauseWeAreStillOnlyHalfWayDoneOrMaybeNotOrMaybeMassiveQueryThatNeverEndsAndExceedsTwoHundredBytesWhichIsTheSizeOfEntrySizeAndShouldThusBeTruncatedAndIamJustGoingToRepeatTheSameCharactersAgainProbablyBecauseWeAreStillOnlyHalfWayDoneOrMaybeNotOrMaybeMassiveQueryThatNeverEndsAndExceedsTwoHundredBytesWhichIsTheSizeOfEntrySizeAndShouldThusBeTruncatedAndIamJustGoingToRepeatTheSameCharactersAgainProbablyBecauseWeAreStillOnlyHalfWayDoneOrMaybeNotOrMaybeMassiveQueryThatNeverEndsAndExceedsTwoHundredBytesWhichIsTheSizeOfEntrySizeAndShouldThusBeTruncatedAndIamJustGoingToRepeatTheSameCharactersAgainProbablyBecauseWeAreStillOnlyHalfWayDoneOrMaybeNotOrMaybeMassiveQueryThatNeverEndsAndExceedsTwoHundredBytesWhichIsTheSizeOfEntrySizeAndShouldThusBeTruncatedAndIamJustGoingToRepeatTheSameCharactersAgainProbablyBecauseWeAreStillOnlyHalfWayDoneOrMaybeNotOrMaybe"
+	queries := []string{
+		"TestQuery",
+		veryLongString,
+		"",
+		"SpecialCharQuery{host=\"2132132\", id=123123}",
+	}
+
+	want := []string{
+		`^{"query":"TestQuery","timestamp_sec":\d+}\x00*,$`,
+		`^{"query":"` + trimStringByBytes(veryLongString, entrySize-40) + `","timestamp_sec":\d+}\x00*,$`,
+		`^{"query":"","timestamp_sec":\d+}\x00*,$`,
+		`^{"query":"SpecialCharQuery{host=\\"2132132\\", id=123123}","timestamp_sec":\d+}\x00*,$`,
+	}
+
+	// Check for inserts of queries.
+	for i := 0; i < 4; i++ {
+		start := 1 + i*entrySize
+		end := start + entrySize
+
+		queryLogger.Insert(context.Background(), queries[i])
+
+		have := string(fileAsBytes[start:end])
+		require.True(t, regexp.MustCompile(want[i]).MatchString(have),
+			"Query not written correctly: %s", queries[i])
+	}
+
+	// Delete first 2 queries
+	for i := 0; i < 2; i++ {
+		queryLogger.Delete(1 + i*entrySize)
+	}
+
+	// Close the query logger
+	require.NoError(t, queryLogger.Close())
```

Comment:
Actually, I think this shouldn't happen in the prod code.
We shouldn't close the engine (which closes the queryLogger) unless we are sure that its consumers have stopped using it.
I think reverting the `cmd/prometheus/main.go` changes in https://github.com/prometheus/prometheus/pull/14064/files is the simplest/safest solution to restore the previous behavior: `logger.Insert` and `logger.Delete` aren't blocked, especially `logger.Delete` but the engine decides to close.

Now, if we believe that `logger.Insert` makes no sense when the instance is shutting down (I'm not aware of any complaints regarding that), we can discuss changing the behavior, but the logic will most likely reside in the engine side, requiring more investigation/tests...

Reply:
> We shouldn't close the engine (which closes the queryLogger) unless we are sure that its consumers have stopped using it.

obviously that would be ideal, but the current code uses deferred calls to delete queries once they complete and that makes it harder to do this

Reply:
I opened https://github.com/prometheus/prometheus/pull/15723